core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java

```diff
@@ -0,0 +1,239 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.endpoint;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_CREATE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_DELETE;
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+import static java.nio.file.StandardWatchEventKinds.OVERFLOW;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Path;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.endpoint.FileWatcherRunnable.FileWatchEvent;
+
+/**
+ * Wraps a {@link WatchService} and allows paths to be registered.
+ */
+final class FileWatcherRegistry implements AutoCloseable {
+
+    /**
+     * Context responsible for watching a {@link FileSystem}. Contains a reference to
+     * a {@link WatchService} and a {@link Thread} which continuously watches for changes
+     * in registered file paths.
+     */
+    static class FileWatchServiceContext {
+
+        private final RestartableThread restartableThread;
+        private final WatchService watchService;
+        private final Map<FileWatchRegisterKey, FileWatchEvent> currWatchEventMap =
+                new ConcurrentHashMap<>();
+
+        /**
+         * Creates a new {@link FileWatchServiceContext}.
+         * @param name the name of the thread used to watch for changes
+         * @param watchService the {@link WatchService} used to monitor for changes
+         */
+        FileWatchServiceContext(String name, WatchService watchService) {
+            restartableThread = new RestartableThread(name, () ->
+                    new FileWatcherRunnable(watchService, this));
+            this.watchService = watchService;
+        }
+
+        /**
+         * Starts to watch changes for the path corresponding to the {@link FileWatchRegisterKey}.
+         * When changes are detected, the {@code callback} is invoked. If the {@link WatchService}
+         * isn't running yet, this method starts a thread to start watching the {@link FileSystem}.
+         * @param watchRegisterKey key that contains the path to be watched
+         * @param callback function invoked on file change
+         *
+         * @throws IllegalArgumentException if failed to locate file or failed to start watching
+         */
+        void register(FileWatchRegisterKey watchRegisterKey, Runnable callback) {
+            final Path dirPath;
+            try {
+                dirPath = watchRegisterKey.filePath().toRealPath().getParent();
```

Comment:
hmmm.. I think it depends on what the user wants -- the user might want to watch the "real file" disregarding symlinks, or intentionally watch symlink directories that may change. I'm starting to think it might actually be better to delegate this decision to the user and not manipulate the input `path`. (especially since I'm not 100% sure what the use-cases for this class will be)
wdyt?

Reply:
Yeah, it'd be simple for the user to call `toRealPath` if they wanted that behavior, so not modifying the input SGTM